external/fv3fit/tests/test_training_loop.py

```diff
@@ -79,3 +80,42 @@ def test_fit_loop_calls_fit(n_epochs):
     with mock_tfdataset_to_tensor_sequence():
         config.fit_loop(mock_model, mock_Xy, validation_data)
     assert mock_model.fit.call_count == 1
+
+
+def test_fit_loop_shuffles_batches():
+    fv3fit.set_random_seed(0)
+    # for test, batch_shuffle_buffer_size should fit all batches
+    # and shuffle_buffer_size should not
+    # in_memory must be False because keras's model.fit handle shuffling for True
+    batch_size = 1000
+    n_samples = 20_000
+    n_batches = n_samples // batch_size
+    config = fv3fit.TrainingLoopConfig(shuffle_buffer_size=batch_size, in_memory=False)
+    mock_model = mock.MagicMock()
+
+    array_x = np.arange(n_samples)
+    x = tf.data.Dataset.from_tensor_slices(array_x).batch(batch_size)
+    y = tf.data.Dataset.from_tensor_slices(array_x).batch(batch_size)
+    data = tf.data.Dataset.zip((x, y))
+
+    config.fit_loop(model=mock_model, Xy=data, validation_data=None, callbacks=())
+    args = mock_model.fit.call_args_list[0]
+    Xy = args.args[0].unbatch()
+    n_samples = sequence_size(Xy)
+    shuffled_x, shuffled_y = next(iter(Xy.batch(n_samples)))
+    np.testing.assert_array_equal(shuffled_x, shuffled_y)
+    assert not np.all(shuffled_x == np.arange(n_samples))
+    # this code is here for manual checking, since it's hard to assert that something
+    # is shuffled
+    # import matplotlib.pyplot as plt
+    # plt.figure()
+    # plt.hist2d(array_x, shuffled_x, bins=n_batches)
+    # plt.xlabel("output index")
+    # plt.ylabel("input index")
+    # plt.show()
+
+    # check that there is some inter-batch shuffling
+    hist, _, _ = np.histogram2d(array_x, shuffled_x, bins=n_batches)
```

Comment:
these tests could be a bit brittle and block future intentional changes in behavior.  I also worry about the runtime..since fv3fit already has a pretty slow test suite. I expect these assertions will survive longer through future refactors if you included them in a unit test rather than this training integration test. 

If there is no such unit to test, the perhaps one should be written since shuffling seems like a separate concern from other aspects of loop fitting.  Can you inject a shuffle implementation into `TrainingLoopConfig`? This would remove the concern of shuffling from `fit_loop`, and allow it to evolve freely/change its shuffling strategy.

Reply:
I'll split the shuffle and rebatching into its own helper function, and add a unit test of that function.

The test is quite fast, actually (2s).

I'd like to VC on changes past that - I intended for these assertions to be things that, at least presently, a user should really think about changing TrainingLoopConfig in a way that breaks the assertions. I think we can compartmentalize this logic into a new swappable component when we need to, but there is significant overhead and complexity when we add swappable components (like Hyperparameters), and I don't think we should do it prematurely. Moving this code to a function like you suggested will at least help keep this as an easy-to-change thing down the line.

Reply:
sounds good for now.